stubs/ext-ds.php

```diff
@@ -0,0 +1,699 @@
+	 */
+	public function copy(): Collection;
+
+	/**
+	 * @return array<TValue>
+	 */
+	public function toArray(): array;
+}
+
+/**
+ * @template TValue
+ * @implements Sequence<TValue>
+ */
+final class Deque implements IteratorAggregate, ArrayAccess, Sequence
+{
+	/**
+	 * @return Deque<TValue>
+	 */
+	public function copy(): Deque
+	{
+	}
+
+	/**
+	 * @param iterable<TValue>|null $values
+	 */
+	public function __construct($values = null)
+	{
+	}
+
+	/**
+	 * @param iterable<TValue> $values
+	 * @return Deque<TValue>
+	 */
+	public function merge($values): Deque
+	{
+	}
+
+	/**
+	 * @param (callable(TValue): bool)|null $callback
+	 * @return Deque<TValue>
+	 */
+	public function filter(callable $callback = null): Deque
+	{
+	}
+
+	/**
+	 * @template TNewValue
+	 * @param callable(TValue): TNewValue $callback
+	 * @return Deque<TNewValue>
+	 */
+	public function map(callable $callback): Deque
+	{
+	}
+
+	/**
+	 * @return Deque<TValue>
+	 */
+	public function reversed(): Deque
+	{
+	}
+
+	/**
+	 * @return Deque<TValue>
+	 */
+	public function slice(int $offset, ?int $length = null): Deque
+	{
+	}
+}
+
+/**
+ * @template TKey
+ * @template TValue
+ *
+ * @implements Collection<TValue>
+ * @implements ArrayAccess<TKey, TValue>
+ * @implements IteratorAggregate<TKey, TValue>
+ */
+final class Map implements IteratorAggregate, ArrayAccess, Collection
+{
+	/**
+	 * @return Map<TKey, TValue>
+	 */
+	public function copy(): Map
+	{
+	}
+
+	/**
+	 * @param iterable<TKey, TValue>|null $values
+	 */
+	public function __construct($values = null)
+	{
+	}
+
+	/**
+	 * @param callable(TKey, TValue): TValue $callback
+	 */
+	public function apply(callable $callback)
+	{
+	}
+
+	/**
+	 * @return Pair<TKey, TValue>
+	 * @throws UnderflowException
+	 */
+	public function first(): Pair
+	{
+	}
+
+	/**
+	 * @return Pair<TKey, TValue>
+	 * @throws UnderflowException
+	 */
+	public function last(): Pair
+	{
+	}
+
+	/**
+	 * @return Pair<TKey, TValue>
+	 * @throws OutOfRangeException
+	 */
+	public function skip(int $position): Pair
+	{
+	}
+
+	/**
+	 * @param iterable<TKey, TValue> $values
+	 * @return Map<TKey, TValue>
+	 */
+	public function merge($values): Map
+	{
+	}
+
+	/**
+	 * @param Map<TKey, TValue> $map
+	 * @return Map<TKey, TValue>
+	 */
+	public function intersect(Map $map): Map
+	{
+	}
+
+	/**
+	 * @param Map<TKey, TValue> $map
+	 * @return Map<TKey, TValue>
+	 */
+	public function diff(Map $map): Map
+	{
+	}
+
+	/**
+	 * @param TKey $key
+	 */
+	public function hasKey($key): bool
+	{
+	}
+
+	/**
+	 * @param TValue $value
+	 */
+	public function hasValue($value): bool
+	{
+	}
+
+	/**
+	 * @param (callable(TKey, TValue): bool)|null $callback
+	 * @return Map<TKey, TValue>
+	 */
+	public function filter(callable $callback = null): Map
+	{
+	}
+
+	/**
+	 * @template TDefault
+	 * @param TKey $key
+	 * @param TDefault $default
+	 * @return TValue|TDefault
+	 * @throws OutOfBoundsException
+	 */
+	public function get($key, $default = null)
```

Comment:
How do I properly annotate this? The method has two different signatures:

If `$default` is given then it is returned if the key does not exist.
If `$default` is not given then TValue is always returned (throws on nonexistent key).

Reply:
This cannot be solved with generics but needs a dynamic return type extension.